File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        cur = root

        while cur or stack:
            while cur:
                # Keep going left
                stack.append(cur)
                cur = cur.left
            
            cur = stack.pop()
            k -= 1

            if k == 0:
                return cur.val

            ## Go to right node
            cur = cur.right

        return root.val

        

File: Data Structures & Algorithms/same-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // Traverse two tree at the same time, and check if each value the same
        
        if ((p == null && q != null) || (q == null && p != null)) {
            return false;
        }

        if (p == null && q == null) {
            return true;
        }
        
        if (p.val != q.val) {
            return false;
        }

        boolean left = isSameTree(p.left, q.left);
        if (left == false){
            return false;
        }
        boolean right = isSameTree(p.right, q.right);
        if (right == false){
            return false;
        }


        return true;
    }
}
